util/src/main/java/com/coderanch/util/cli/InputUtility.java

```diff
@@ -0,0 +1,230 @@
+package com.coderanch.util.cli;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collector;
+
+/**
+ * Multipurpose input utility for use with a command line interface.
+ * The user can pass a prompt and optional validation.
+ * The utility will read the provided input stream and return a possibly validated String or primitive.
+ */
+public class InputUtility {
+
+    private static final List<String> YES = Collections.unmodifiableList(List.of("y", "yes"));
+    private static final List<String> NO = Collections.unmodifiableList(List.of("n",  "no"));
+
+    private InputStream inputStream;
+
+    /**
+     * If no input stream is provided, System.in will be used.
+     */
+    public InputUtility() {
+        this(System.in);
+    }
```

Comment:
Personally I think the input stream should always be explicitly specified.

Reply:
But it's nice to have a reasonable default, isn't it?  Especially since most of the time you'll be using System.in, except for testing.

Reply:
I agree with reasonable defaults, as long as reasonable defaults don't lead to environment-specific behavior.

`System.in` is really not much more than a mutable global variable and its behavior depends on operating system, execution profile, and more.

I like to compare it to `String.getBytes()`. This method is deprecated because it uses an environment-specific character set.

Anyway, this is not a blocker for me, if you and Al want to keep it in I won't reject the PR because of it.